runtime/src/storage/mkvs/sync/proof.rs

```diff
@@ -61,6 +64,112 @@ pub struct Proof {
     pub entries: Vec<Option<RawProofEntry>>,
 }
 
+struct ProofNode {
+    serialized: Vec<u8>,
+    children: Vec<Hash>,
+}
+
+/// A Merkle proof builder.
+pub struct ProofBuilder {
+    root: Hash,
+    included: BTreeMap<Hash, ProofNode>,
+}
+
+impl ProofBuilder {
+    /// Create a new proof builder for the given root hash.
+    pub fn new(root: Hash) -> Self {
+        Self {
+            root,
+            included: BTreeMap::new(),
+        }
+    }
+
+    /// Add a node to the set of included nodes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the node is not clean.
+    pub fn include(&mut self, node: &NodeBox) {
+        if !node.is_clean() {
+            panic!("proof builder: node is not clean");
+        }
+
+        // If node is already included, skip.
+        let nh = node.get_hash();
+        if self.included.contains_key(&nh) {
+            return;
+        }
+        let mut pn = ProofNode {
+            serialized: node
+                .compact_marshal_binary()
+                .expect("marshaling node in proof"),
+            children: vec![],
+        };
+
+        // For internal nodes, also include children.
+        if let NodeBox::Internal(nd) = node {
+            let ch = nd
+                .left
+                .borrow()
+                .node
+                .as_ref()
+                .map(|n| n.borrow().get_hash())
+                .unwrap_or(Hash::empty_hash());
+            pn.children.push(ch);
+
+            let ch = nd
+                .right
+                .borrow()
+                .node
+                .as_ref()
+                .map(|n| n.borrow().get_hash())
+                .unwrap_or(Hash::empty_hash());
+            pn.children.push(ch);
+        }
+
+        self.included.insert(nh, pn);
+    }
+
+    /// Build the proof.
```

Comment:
```suggestion
    /// Build unverified proof.
```
Not sure if it is obvious to the caller that `build` can construct invalid proofs. I would update the comment or maybe mention somewhere that the caller is responsible for creating valid proofs.